File: 백준/Silver/2805. 나무 자르기/나무 자르기.c
#include<stdio.h>
long long tree[1000001];

long long getTree(int s, long long e, int N, long long M){
	long long left = s, right=e, mid, sum;
	long long upper_bnd = -1;
	
	while(left <= right){
		mid = (left + right) / 2;
		
		sum=0;
		for(int i=0; i<N; i++){
			if(tree[i] > mid)
				sum += (tree[i] - mid);
		}
		
		if(sum >= M){
			upper_bnd = mid;
			left = mid +1;
		}
		else
			right = mid -1;
	}
	return upper_bnd;
}
int main(){
	int N;
    long long M, max;

	scanf("%d %lld", &N, &M);
	max=0;
	for(int i=0; i<N; i++){
		scanf("%lld", &tree[i]);
		
		if(max < tree[i]){
			max = tree[i];
		}
		
	}
	
	printf("%lld\n", getTree(0, max, N, M));
	return 0;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.util.*;
import java.io.*;
public class Main {

	static class FastReader {
		BufferedReader br;
		StringTokenizer st;
		
		public FastReader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}
		
		String next() {
		
			while(st == null || !st.hasMoreTokens()) {
				try{
					st = new StringTokenizer(br.readLine());
				} catch(IOException e) {
					e.printStackTrace();
				}
			}
			
			return st.nextToken();
		}
		
		long nextLong() {
		
			return Long.parseLong(next());
		}
	}
	
	static StringBuilder sb = new StringBuilder();
	
	static long N, M;
	static long[] trees;
	
	static void input() {
	
		FastReader scan = new FastReader();
		N = scan.nextLong();
		M = scan.nextLong();
		trees = new long[(int)N+1];
		
		for(int i=1; i<N+1; i++) {
			trees[i] = scan.nextLong();
		}
	}
	
	static boolean determination(long H) {
		
		long sum = 0;
		
		for(int i=1; i<N+1; i++) {
			if(trees[i] > H) {
				sum += (long) trees[i] - H;
			}
		}
		
		return sum >= M;
	}
	
	static void solution() {
	
		long left = 0, right = 2000000000, answer = 0;
		
		while(left <= right) {
		
			long mid = (left+right) / 2;
			if(determination(mid)) {
				answer = mid;
				left = mid + 1;
			} else {
				right = mid - 1;
			}
		}
		
		System.out.println(answer);
	}
	
	public static void main(String args[]) {
	
		input();
		solution();
	}
}